src/dbnode/storage/shard.go

```
@@ -2660,74 +2661,217 @@ func (s *dbShard) Repair(
+				Namespace:   sourceNsID,
+				Shard:       sourceShardID,
+				BlockStart:  sourceBlockVolume.blockStart,
+				VolumeIndex: sourceBlockVolume.latestVolume,
+			},
+			FileSetType:      persist.FileSetFlushType,
+			StreamingEnabled: true,
+		}
+
+		if err := blockReader.Open(openOpts); err != nil {
+			s.logger.Error("blockReader.Open", zap.Error(err))
+			return 0, err
+		}
+		if blockReader.Entries() > maxEntries {
+			maxEntries = blockReader.Entries()
+		}
+
+		blockReadersToClose = append(blockReadersToClose, blockReader)
+	}
+
+	crossBlockReader, err := fs.NewCrossBlockReader(blockReaders, s.opts.InstrumentOptions())
 	if err != nil {
+		s.logger.Error("NewCrossBlockReader", zap.Error(err))
 		return 0, err
 	}
+	defer crossBlockReader.Close()
 
-	openOpts := fs.DataReaderOpenOptions{
-		Identifier: fs.FileSetFileIdentifier{
-			Namespace:   sourceNsID,
-			Shard:       sourceShard.ID(),
-			BlockStart:  sourceBlockStart,
-			VolumeIndex: latestSourceVolume,
+	tileOpts := tile.Options{
+		FrameSize:          xtime.UnixNano(opts.Step.Nanoseconds()),
+		Start:              xtime.ToUnixNano(opts.Start),
+		ReaderIteratorPool: s.opts.ReaderIteratorPool(),
+	}
+
+	// TODO: these should probably be pooled
+	readerIter, err := tile.NewSeriesBlockIterator(crossBlockReader, tileOpts)
+	if err != nil {
+		s.logger.Error("error when creating new series block iterator", zap.Error(err))
+		return 0, err
+	}
+
+	closed := false
+	defer func() {
+		if !closed {
+			if err := readerIter.Close(); err != nil {
+				// NB: log the error on ungraceful exit.
+				s.logger.Error("could not close read iterator on error", zap.Error(err))
+			}
+		}
+	}()
+
+	encoder := s.opts.EncoderPool().Get()
+	defer encoder.Close()
+	encoder.Reset(opts.Start, 0, targetSchemaDescr)
+
+	latestTargetVolume, err := s.LatestVolume(opts.Start)
+	nextVersion := latestTargetVolume + 1
+	if err != nil {
+		s.logger.Error("latestTargetVolume", zap.Error(err))
+		nextVersion = 0
```

This should return an error, not just set nextVersion to zero. This is definitely unexpected, we should always be able to resolve the next version unless something is really messed up (and in which case better to return error).

---

👍